Add compatibility tests for ConvertTo-Json

yotsuda opened the pull request:
# PR Summary

Add compatibility tests for `ConvertTo-Json` to improve test coverage.

## PR Context

This PR adds tests to validate `ConvertTo-Json` behavior, extracted from #26637 as suggested by @iSazonov.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] These changes were tested
- [x] Test added (test-only PR)

## Tests Added

The following tests are added to `ConvertTo-Json.Tests.ps1`:

| Test | Description |
|------|-------------|
| Uri serialization | Validates Uri objects serialize correctly |
| Enum default | Validates enums serialize as numbers by default |
| Enum -EnumsAsStrings | Validates -EnumsAsStrings parameter |
| null | Validates null serialization |
| Arrays | Validates array serialization |
| Hashtable | Validates ordered hashtable serialization |
| Nested objects | Validates nested PSCustomObject serialization |
| Default escaping | Validates no escaping by default |
| -EscapeHandling EscapeHtml | Validates HTML character escaping |
| -EscapeHandling EscapeNonAscii | Validates non-ASCII escaping |
| -Depth | Validates Depth parameter |
| -AsArray | Validates AsArray parameter |
| Multiple pipeline objects | Validates array output for multiple inputs |
| Depth over 100 | Validates depth limit enforcement |

yotsuda committed:
Add compatibility tests for ConvertTo-Json

iSazonov commented:
I guess we can add more tests while working on V2 JSON serializer. So I will keep this open some time.

iSazonov commented:
If there are duplicates we could simply add comments in existing tests to describe additional scenarios are tested. If we can not create clear comments please keep duplicate tests.

yotsuda committed:
Refactor ConvertTo-Json tests to remove duplicates and improve coverage

yotsuda commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1:
```diff
@@ -156,4 +156,84 @@ Describe 'ConvertTo-Json' -tags "CI" {
+    It 'Multiple objects from pipeline should be serialized as array' {
+        $json = 1, 2, 3 | ConvertTo-Json -Compress
+        $json | Should -BeExactly '[1,2,3]'
+    }
```
Removed. Covered by [lines 75-76](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1#L75-L76).

yotsuda commented on test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1:
```diff
@@ -156,4 +156,84 @@ Describe 'ConvertTo-Json' -tags "CI" {
+
+    It 'Should serialize null correctly' {
+        $null | ConvertTo-Json | Should -BeExactly 'null'
+    }
```
Removed. Covered by [line 96](https://github.com/PowerShell/PowerShell/blob/master/test/powershell/Modules/Microsoft.PowerShell.Utility/ConvertTo-Json.Tests.ps1#L96).

yotsuda committed:
Add nested raw object serialization tests

yotsuda committed:
Add STJ scalar consistency, PSObject distinction, and depth warning tests

yotsuda committed:
Add test for array with ETS properties serialization

iSazonov commented:
@yotsuda  Have we tests to cover all code paths for `if (currentDepth > context.MaxDepth)`? Notice, we don't add extended/adapted properties if isPurePSObj = true; and add otherwise.

yotsuda committed:
Add tests for depth exceeded behavior with PSObject ETS properties

yotsuda committed:
Fix DateTime ETS test to use UTC and add missing Add-Member

iSazonov commented:
@yotsuda Thanks! I see if object is pure PSObject LanguagePrimitives.ConvertTo() does all work and add ETS properties so we skipp AddPSProperty(). But I wonder why these properties are not added if isPurePSObj == false - I'd expect AddPSProperty() should add. Thoughts?

yotsuda committed:
Add test for non-pure PSObject with ETS properties at depth exceeded

yotsuda commented:
@iSazonov Let me trace through the code flow (referring to V1/JsonObject.cs).

When `isPurePSObj == false` and `currentDepth > MaxDepth`:

1. The object is converted to string via `LanguagePrimitives.ConvertTo(obj, typeof(string), ...)` ([line 618-619](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L618-L619))
2. `AddPsProperties` is called with the converted string
3. Since `isPurePSObj == false`, we don't early-return at [line 671-673](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L671-L673)
4. The string is wrapped in a new Dictionary with key "value" ([line 680-682](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L680-L682))
5. `AppendPsProperties` is called ([line 685](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L685))

In `AppendPsProperties`:
- [Line 710-712](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L710-L712) checks if the **original** `psObj.BaseObject` is string or DateTime - if so, it returns early (no ETS properties added)
- [Line 716-719](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L716-L719): Since `isCustomObj == false`, only Extended properties are searched (not Adapted)
- Any Extended properties found are added to the dictionary

After `AppendPsProperties` returns:
- [Line 687-690](https://github.com/PowerShell/PowerShell/blob/eb2ead439/src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/JsonObject.cs#L687-L690): If the original `rv` was not a dictionary and no properties were added (dict.Count == 1), the original string is returned instead of the dictionary wrapper

So the ETS properties **are** added for `isPurePSObj == false` when:
- The original BaseObject is not string/DateTime, AND
- Extended properties exist on the PSObject

The properties are **not** added when:
- BaseObject is string/DateTime (intentional, these types use their natural serialization)
- No Extended properties exist

---

**Note for #26654 (V2/STJ):** While testing this, I found an issue in `WriteProperty` where `MaxDepth == 0` directly called `ToString()` without going through `WriteDepthExceeded`, so ETS properties on nested PSObjects were not included. I've fixed this in #26654 (commit 67a1bbd40) and added a test to cover this case.

yotsuda committed:
Add test for Uri with ETS properties

yotsuda committed:
Fix cross-platform test failures for FileInfo serialization
